### PR TITLE
Support reverse scan (Closes #111)

### LINKED ISSUE #111: Add support for backward scanning
Hello, I was wondering if there's a plan to add support for backward scanning for Hbase 0.98 / CDH5?  Thanks!

### PR DESCRIPTION
**Pull request by HeartSaVioR**
First of all, this PR is based on @jnfang (Jiayun Fang) reverse scan implementation.

From #111, I mentioned @caprius and @jnfang to persuade her to make her work finished.
But no one responded, and I mailed her (Jiayun Fang) via commit mail address but she also doesn't respond, too.

I've signed CLA, but she may not sign CLA.
If this PR can't be pulled because of license issue, I have a mind to learn details of implementation and re-implement.
Just I'm inspired on her great work, so I'd like to preserve her commits to make sure she give a credit.

Please feel free to let me know if it makes issue.

And I had a hard time to cherry-pick each her commits to ```next``` branch, so I gave up and took her branch, and upmerge ```next``` branch to merge.

As @tsuna adds "feature" label to #111, I'd like to get reviewed from @tsuna, and other collaborators / owners.

Thanks in advance!

ps. Some integration tests fails on my machine but these are only about write_time, write_time exceeds 5 secs.

### CODE REVIEW

**Comment by caprius** [comment]
Hi @HeartSaVioR, Apologies for the delay in responding.  We are currently in the process of testing these changes in a production system before merging the code.

And, you are correct that it's hard to cherry-pick these commits to the next branch because further changes were made to that branch that need to be merged with our changes.  I have done this internally, but have not pushed the changes to @jnfang's branch.

Although I can't guarantee the timeline, our current plan is to have it in production within a month, at which point, we would feel comfortable merging our changes.  Does that work for you?

**Comment by HeartSaVioR** [comment]
Hello @caprius and @jnfang , 

Great news that you are about to finish the work!
I'm fine to wait a month or more, but I need to know when it will be done approximately because I am currently working on it and I'd like to share the plan to our team.
If it requires some months, I'll have to re-work and send another pull request.

Btw, as you can see the changes in this PR, I'm working not only upmerging, but also refactoring to make it beauty and consistent.
I think it might help to your work, but not sure cause you don't publish current version of your work.
As my work is based on your work, I have a mind to contribute your work.
If you want to co-work, please feel free to let me know.

Thanks for the amazing work!

**Comment by HeartSaVioR** [comment]
If you want me to not using your work (for various reasons) and wait your work to finish, please feel free to let me know, too. :)

**Comment by HeartSaVioR** [comment]
@caprius Kindly reminder: Any updates on this issue?

**Comment by caprius** [comment]
Hi @HeartSaVioR, we tried deploying to production a few weeks ago, but ran into some problems due to a known bug in asynchbase (https://issues.apache.org/jira/browse/HBASE-14283), which we've just received the patch for.  The updated plan is to roll it out again this week.  

As an aside, you should also plan to apply the patch if your version of Hbase or CDH doesn't include it.  Without it, we saw requests occasionally failing.

As I mentioned before, I made a few changes to the repo that I haven't yet merged.  I will merge those changes by tomorrow so that you're free to use our work and add to it.  We will still wait for the service to be stable in production for some time before upstreaming our changes, but you are welcome to use our work in the meantime.

**Comment by HeartSaVioR** [comment]
@caprius 
For clarification, you said it is a known bug in asyncbase but you posted an HBase issue. It seems to be HBase issue, right?
And thanks for letting me know the current progress status.

Please let me know when you push your changes to your repo if you don't feel bothered.
Hope to see you and @jnfang 's great work to the master of Asyncbase.

**Comment by caprius** [comment]
@HeartSaVioR Ah apologies, I misspoke.  You are correct, a bug in Hbase.  If you don't mind me asking, what are your current plans with reverse fetch, in terms of changes you're making and your timeline?

**Comment by HeartSaVioR** [comment]
@caprius Yeah, fine. :)
We're not using reverse scan and don't have actual plan to use, just wanted to have this feature.
When it's merged into master and released to official release version of Asynchbase, we'll implement some ideas which rely on reverse scan. (it means those ideas are not mandatory for now.)

**Comment by caprius** [comment]
Ah gotcha.  A quick update.  I'm still working on pushing up my changes.
I'm having a couple difficulties, but I'll let you know as soon as it's up.

**Comment by HeartSaVioR** [comment]
Hi @caprius,
our team is interested in this feature now, so we just build custom version of asynchbase based on this pull request.
But we still want your update (in order to update our custom version), and hope to merge into next branch (finally master).

Please let me know whenever you have any news. Thanks!

**Comment by HeartSaVioR** [comment]
@caprius Kindly reminder: Any updates on this issue?

**Comment by manolama** [comment]
@jnfang @caprius I have need of this as well and would love to merge it in, just need the CLA's signed then I can help clean up a bit. Thanks!

File: src/Bytes.java [hunk 1]
```diff
@@ -50,6 +50,19 @@
...
   }
 
+  /**
+   * Create a max byte array with the specified max byte count
+   * @param num_bytes the length of returned byte array
+   * @return the created max byte array
+   */
+  public static byte[] createMaxByteArray(int num_bytes) {
+    byte[] max_byte_arr = new byte[num_bytes];
+    for (int i = 0; i < max_byte_arr.length; i++) {
```
Comment (manolama) [inline src/Bytes.java]: Could just use ``Arrays.fill(max_byte_arr, (byte) 0xff);`` here

File: src/RegionInfo.java [hunk 2]
```diff
@@ -225,6 +225,46 @@ static RegionInfo fromKeyValue(final KeyValue kv,
     return Arrays.copyOf(region_name, comma);
   }
 
+  /**
+   * Given name of a region, returns its start key
+   * @throws IllegalArgumentException if the name of the region is malformed
+   * @param region_name Full region_name created in the constructor
+   * @return byte Array of the start key
+   */
+  static byte[] startKeyFromRegionName(final byte[] region_name){
```
Comment (manolama) [inline src/RegionInfo.java]: Looks like this can just be a member function instead of static.

File: src/Scanner.java [hunk 3]
```diff
@@ -1188,6 +1282,7 @@ private int predictSerializedSize() {
       size += start_key.length;  // The start key.
       size += 3;  // vint: stop key length (3 bytes => max length = 32768).
       size += stop_key.length;  // The stop key.
+      size += 1;  // bool: Whether the scanner is in reverse
```
Comment (manolama) [inline src/Scanner.java]: Nit: Don't need this hear since pre-protobuf versions don't support reverse scanning :)

File: src/Scanner.java [hunk 4]
```diff
@@ -205,6 +209,41 @@
...
+  *
+  * In a reversed scan, the first Scanner opened is opened on the region
+  * that the Scanner start key is in (the same as forward scan). But 
+  * subsequent scanners are opened via the overloaded 
+  * {@link #OpenScannerRequest} and the region to be opened on is 
+  * found through a META lookup using
+   * {@link HBaseClient#locateRegionClosestBeforeKey}.
+  * 
+  */
+  public void setReverse(){
```
Comment (manolama) [inline src/Scanner.java]: Let's change this to `setReversed(boolean reversed)`` to match the OS API.

File: src/Scanner.java [hunk 5]
```diff
@@ -205,6 +209,41 @@
   }
 
   /**
+  * Returns boolean indicating if scanner is in reversed direction
+  */
+  public boolean getReversed(){
+    return is_reversed;
```
Comment (manolama) [inline src/Scanner.java]: Should be ``isReversed()`` to match OS

File: src/Bytes.java (hunk 1 above)
Comment (HeartSaVioR) [inline src/Bytes.java]: Great! Will address.

File: src/RegionInfo.java (hunk 2 above)
Comment (HeartSaVioR) [inline src/RegionInfo.java]: This method is just a utility method, and isn't related to any fields / instances of RegionInfo.
(similar kind of RegionInfo#tableFromRegionName())

Maybe we move out these two methods if we know the better place.

File: src/Scanner.java (hunk 3 above)
Comment (HeartSaVioR) [inline src/Scanner.java]: Good, will address.

File: src/Scanner.java (hunk 4 above)
Comment (HeartSaVioR) [inline src/Scanner.java]: Great. Will address.
Actually javadoc is wrong even now, supposing there's some refactoring but javadoc was not fixed.

File: src/Scanner.java (hunk 5 above)
Comment (HeartSaVioR) [inline src/Scanner.java]: Great. Will address.

**Comment by HeartSaVioR** [comment]
@manolama Addressed your review.
There seems to have intermittent failures on integration test.

```
09:06:24.884 [main] ERROR org.hbase.async.TestIntegration - 1 tests failed: [appendTwiceBatchedNoReturnRead(org.hbase.async.TestIntegration): arrays first differed at element [0]; expected:<118> but was:<50>]
```
```
08:54:09.293 [main] ERROR org.hbase.async.TestIntegration - 2 tests failed: [appendTwiceBatchedReturnRead(org.hbase.async.TestIntegration): arrays first differed at element [0]; expected:<118> but was:<50>, incrementCoalescingWithAmountsTooBig(org.hbase.async.TestIntegration): expected:<562949953421312> but was:<281474976710656>]
```

But I can also see that all tests are passed, and failed tests are not related to this PR.

I also see error from unit test which seems odd for me.

```
30) ticketRenewalTaskException(org.hbase.async.auth.TestLogin)
java.lang.LinkageError: loader constraint violation: when resolving overridden method "$javax.security.auth.kerberos.KerberosPrincipal$$EnhancerByMockitoWithCGLIB$$fbe7c767.implies(Ljavax/security/auth/Subject;)Z" the class loader (instance of org/powermock/core/classloader/MockClassLoader) of the current class, $javax/security/auth/kerberos/KerberosPrincipal$$EnhancerByMockitoWithCGLIB$$fbe7c767, and its superclass loader (instance of <bootloader>), have different Class objects for the type javax/security/auth/Subject used in the signature
	at sun.reflect.GeneratedSerializationConstructorAccessor12.newInstance(Unknown Source)
	at java.lang.reflect.Constructor.newInstance(Constructor.java:422)
	at org.objenesis.instantiator.sun.SunReflectionFactoryInstantiator.newInstance(SunReflectionFactoryInstantiator.java:45)
	at org.objenesis.ObjenesisBase.newInstance(ObjenesisBase.java:73)
	at org.mockito.internal.creation.jmock.ClassImposterizer.createProxy(ClassImposterizer.java:128)
	at org.mockito.internal.creation.jmock.ClassImposterizer.imposterise(ClassImposterizer.java:63)
	at org.mockito.internal.creation.jmock.ClassImposterizer.imposterise(ClassImposterizer.java:56)
	at org.mockito.internal.creation.CglibMockMaker.createMock(CglibMockMaker.java:23)
	at org.powermock.api.mockito.internal.mockmaker.PowerMockMaker.createMock(PowerMockMaker.java:43)
	at org.mockito.internal.util.MockUtil.createMock(MockUtil.java:26)
	at org.mockito.internal.MockitoCore.mock(MockitoCore.java:51)
	at org.mockito.Mockito.mock(Mockito.java:1243)
	at org.mockito.Mockito.mock(Mockito.java:1120)
	at org.hbase.async.auth.TestLogin.before(TestLogin.java:120)
	at sun.reflect.GeneratedMethodAccessor1.invoke(Unknown Source)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:497)
	at org.junit.internal.runners.MethodRoadie.runBefores(MethodRoadie.java:132)
	at org.junit.internal.runners.MethodRoadie.runBeforesThenTestThenAfters(MethodRoadie.java:95)
	at org.powermock.modules.junit4.internal.impl.PowerMockJUnit44RunnerDelegateImpl$PowerMockJUnit44MethodRunner.executeTest(PowerMockJUnit44RunnerDelegateImpl.java:294)
	at org.powermock.modules.junit4.internal.impl.PowerMockJUnit47RunnerDelegateImpl$PowerMockJUnit47MethodRunner.executeTestInSuper(PowerMockJUnit47RunnerDelegateImpl.java:127)
	at org.powermock.modules.junit4.internal.impl.PowerMockJUnit47RunnerDelegateImpl$PowerMockJUnit47MethodRunner.executeTest(PowerMockJUnit47RunnerDelegateImpl.java:82)
	at org.powermock.modules.junit4.internal.impl.PowerMockJUnit44RunnerDelegateImpl$PowerMockJUnit44MethodRunner.runBeforesThenTestThenAfters(PowerMockJUnit44RunnerDelegateImpl.java:282)
	at org.junit.internal.runners.MethodRoadie.runTest(MethodRoadie.java:86)
	at org.junit.internal.runners.MethodRoadie.run(MethodRoadie.java:49)
	at org.powermock.modules.junit4.internal.impl.PowerMockJUnit44RunnerDelegateImpl.invokeTestMethod(PowerMockJUnit44RunnerDelegateImpl.java:207)
	at org.powermock.modules.junit4.internal.impl.PowerMockJUnit44RunnerDelegateImpl.runMethods(PowerMockJUnit44RunnerDelegateImpl.java:146)
	at org.powermock.modules.junit4.internal.impl.PowerMockJUnit44RunnerDelegateImpl$1.run(PowerMockJUnit44RunnerDelegateImpl.java:120)
	at org.junit.internal.runners.ClassRoadie.runUnprotected(ClassRoadie.java:33)
	at org.junit.internal.runners.ClassRoadie.runProtected(ClassRoadie.java:45)
	at org.powermock.modules.junit4.internal.impl.PowerMockJUnit44RunnerDelegateImpl.run(PowerMockJUnit44RunnerDelegateImpl.java:118)
	at org.powermock.modules.junit4.common.internal.impl.JUnit4TestSuiteChunkerImpl.run(JUnit4TestSuiteChunkerImpl.java:104)
	at org.powermock.modules.junit4.common.internal.impl.AbstractCommonPowerMockRunner.run(AbstractCommonPowerMockRunner.java:53)
	at org.powermock.modules.junit4.PowerMockRunner.run(PowerMockRunner.java:53)
	at org.junit.runners.Suite.runChild(Suite.java:127)
	at org.junit.runners.Suite.runChild(Suite.java:26)
	at org.junit.runners.ParentRunner$3.run(ParentRunner.java:238)
	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:63)
	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:236)
	at org.junit.runners.ParentRunner.access$000(ParentRunner.java:53)
	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:229)
	at org.junit.runners.ParentRunner.run(ParentRunner.java:309)
	at org.junit.runner.JUnitCore.run(JUnitCore.java:160)
	at org.junit.runner.JUnitCore.run(JUnitCore.java:138)
	at org.junit.runner.JUnitCore.run(JUnitCore.java:117)
	at org.junit.runner.JUnitCore.runMain(JUnitCore.java:96)
	at org.junit.runner.JUnitCore.runMainAndExit(JUnitCore.java:47)
	at org.junit.runner.JUnitCore.main(JUnitCore.java:40)
```

Since I don't have any idea why it happens, so I'd like to ask you a favor of checking this.
I'm using OSX 10.11.3 and java version is here.

```
java version "1.8.0_66"
Java(TM) SE Runtime Environment (build 1.8.0_66-b17)
Java HotSpot(TM) 64-Bit Server VM (build 25.66-b17, mixed mode)
```

**Comment by HeartSaVioR** [comment]
@manolama 
Travis CI is failing due to JVM crash on OpenJDK 6. (buffer overflow detected) 
Oracle JDK 7 is passed.

**Comment by manolama** [comment]
Awesome, thanks for taking a stab at that @HeartSaVioR. I had noticed one integration wasn't passing against HBase 1.0.0 so I'll try your updates then see if I get the intermittent failures. Also would you mind rebasing against master and merging into a single commit? Thanks!

**Comment by HeartSaVioR** [comment]
@manolama 
I ran integration test with PredictionIO which consists HBase 1.0.0 as standalone (non-distributed).

And I don't get what you mean about master. Do you mean `origin` (in point of repo's view)?

And I'm curious that merging this into a single commit preserve credit for @jnfang. If it's not, it would be beyond the CLA issue.

Other way I'll spend some amount of time to reimplement this and get rid of CLA issue with @jnfang.

**Comment by jnfang** [comment]
Thank you for reviewing @manolama! @HeartSaVioR, I would appreciate credit if you are thinking of merging it in but I am unable to work on this for a few months. I'm going to loop in @caprius who also has made changes since this code I believe.

**Comment by HeartSaVioR** [comment]
Hi, @jnfang.
Why I mention only you as preserving credit is that this PR is consists of commits which authors are only you and me. So, in order to merge this in, we need only your CLA. 
I'll take care of all the following works with @manolama so you don't need to work any further.

Background: 
This PR is submitted Oct 2015, and I was waiting you and @caprius more than 4 months to show up another complete version of the PR, but it didn't happen. 
I'm sorry @caprius but I don't want to wait any longer and want to take my credit that refines @jnfang  work to be merged in as `next` branch right now.

**Comment by HeartSaVioR** [comment]
Even though I wrote a mail to @tsuna privately to review this PR. 
That's what I was doing for this PR to be merged into origin repo. This reflects my will to introduce reverse scan into origin repo. of AsyncHBase.

**Comment by HeartSaVioR** [comment]
@manolama 
If @jnfang signs CLA and want to take her credit from this PR, I suggest that we can squash this PR into two commits instead of one, which one is for commits which author is @jnfang, and another one is for commits which author is me.
Does it still make sense?

**Comment by manolama** [comment]
Thanks @jnfang! If you and @caprius could fill out the https://docs.google.com/forms/d/1VVqKFTYNp3btgwXE51Zf9_Sfq66yZ4JbuEFrLlTD70E/viewform?embedded=true&formkey=dFNiOFROLXJBbFBmMkQtb1hNMWhUUnc6MQ really quick then we can get it in. I'll be happy to finish up testing and any other bits that may be needed.

@HeartSaVioR Yeah, that'd work great to flatten into two commits. I try avoiding having tons of tiny individual commits as it makes it a little harder to track changes.

**Comment by HeartSaVioR** [comment]
@manolama 
I gave a try and found it takes amount of works since there're so many conflicts during rebasing so stopped working on this for now.
(I even think about rework since I feel it would be not very much harder than rebase...)
I'd like to see completed CLA signing first, and give it a try again.

ps. Maybe I'm missing here, but I don't see a reason why it needs @caprius's CLA to merge this in.

**Comment by caprius** [comment]
@HeartSaVioR thanks for your work on merging and improving this code!  Apologies for not committing my code before.  And, I think you are correct in that only @jnfang will need to sign the CLA.

**Comment by jnfang** [comment]
Just signed the CLA!

**Comment by HeartSaVioR** [comment]
@jnfang 
Thanks for signing CLA quickly! 
Btw, seems like jfang@box.com is not associated to your github account. You may want to associate that email address to let commits applied to your github account. 

@manolama 
I've done with rebase. 
Since PR contains commit for release note by @tsuna so I squashed commits into three, each commits are crafted by same author.

Please take a look and let me know if it requires any further works. Thanks!

**Comment by jnfang** [comment]
I actually no longer have access to jfang@box.com. Will the commit still show up on my github account @HeartSaVioR ?

**Comment by HeartSaVioR** [comment]
@jnfang 
I don't remember that clearly, but when you link additional email account to github account, verification mail is sent to that email. So, you may need to ask Github to link accounts manually.

**Comment by HeartSaVioR** [comment]
@manolama Ping

**Comment by HeartSaVioR** [comment]
@manolama 
Kindly reminder: I'd love to see when you're available to guide for further works on merging, or go merging.

**Comment by manolama** [comment]
Merged in 1.8.0-SNAPSHOT in f6a8ccb7e55ed9bc0aad265345da4c679e750055. Thanks!

**Comment by HeartSaVioR** [comment]
Thanks @manolama!

**Pull request closed without merging**